### PR TITLE
fix: make `fixes` and `refs` detection stricter

### PR DESCRIPTION
**Pull request by richardlau**
Only parse `fixes` and `refs` links at the beginning of lines.

Previously the parser was picking up the "fix" type Conventional
Commits titles from the changelog in body of the npm update pull
requests.

Refs: https://github.com/nodejs/node/pull/42382#issuecomment-1072920361

cc @Trott 

---

I have updated one of the existing test cases as otherwise the test was failing with this change. This test was from https://github.com/nodejs/node-core-utils/issues/117, for https://github.com/nodejs/node/pull/17107. I ran `node ../node-core-utils/bin/get-metadata.js 17107` with this diff:
```diff
diff --git a/lib/links.js b/lib/links.js
index 4d313a1..639ced5 100644
--- a/lib/links.js
+++ b/lib/links.js
@@ -14,6 +14,7 @@ export class LinkParser {
     this.owner = owner;
     this.repo = repo;
     this.$ = cheerio.load(html);
+    console.log(JSON.stringify(html));
   }
 
   getFixesUrlsFromArray(arr) {
```
and found the HTML differs from what is in the test fixture, so I updated it. The test passes was the updated HTML. I used the same diff to capture the test data for https://github.com/nodejs/node/pull/42382.

### CODE REVIEW

**Review by Trott — APPROVED** [review]

File: lib/links.js [hunk 1]
```diff
@@ -1,8 +1,8 @@
 import cheerio from 'cheerio';
 
-const FIXES_RE = /(Close[ds]?|Fix(e[ds])?|Resolve[sd]?)\s*:\s*(\S+)/mgi;
+const FIXES_RE = /^(Close[ds]?|Fix(e[ds])?|Resolve[sd]?)\s*:\s*(\S+)/mgi;
```
Comment (tniessen) [inline lib/links.js:3]:
```suggestion
const FIXES_RE = /^\s*(Close[ds]?|Fix(e[ds])?|Resolve[sd]?)\s*:\s*(\S+)/mgi;
```

File: lib/links.js [hunk 2]
```diff
@@ -1,8 +1,8 @@
...
 const FIX_RE = /(Close[ds]?|Fix(e[ds])?|Resolve[sd]?)\s*:\s*(\S+)/i;
-const REFS_RE = /Refs?\s*:\s*(\S+)/mgi;
+const REFS_RE = /^Refs?\s*:\s*(\S+)/mgi;
```
Comment (tniessen) [inline lib/links.js:5]:
```suggestion
const REFS_RE = /^\s*Refs?\s*:\s*(\S+)/mgi;
```

**Review by tniessen — APPROVED** [review]

File: lib/links.js [hunk 3]
```diff
@@ -1,8 +1,8 @@
 import cheerio from 'cheerio';
 
-const FIXES_RE = /(Close[ds]?|Fix(e[ds])?|Resolve[sd]?)\s*:\s*(\S+)/mgi;
+const FIXES_RE = /^\s*(Close[ds]?|Fix(e[ds])?|Resolve[sd]?)\s*:\s*(\S+)/mgi;
```
Comment (Trott) [inline lib/links.js:3]: By tolerating spaces at the start of the line, might this then match the "Fixes:" etc. in patches we float from V8 etc.? (Maybe that's not a problem?)

Comment (Trott) [reply lib/links.js:3]: Maybe for a separate PR, but I think we probably don't need to tolerate spaces before the colon:

```suggestion
const FIXES_RE = /^\s*(Close[ds]?|Fix(e[ds])?|Resolve[sd]?):\s*(\S+)/mgi;
```

File: lib/links.js [hunk 4]
```diff
@@ -1,8 +1,8 @@
...
 const FIX_RE = /(Close[ds]?|Fix(e[ds])?|Resolve[sd]?)\s*:\s*(\S+)/i;
-const REFS_RE = /Refs?\s*:\s*(\S+)/mgi;
+const REFS_RE = /^\s*Refs?\s*:\s*(\S+)/mgi;
```
Comment (Trott) [inline lib/links.js:5]: Maybe for a separate PR, but I think we probably don't need to tolerate spaces before the colon:

```suggestion
const REFS_RE = /^\s*Refs?:\s*(\S+)/mgi;
```

File: lib/links.js (hunk 3 above)
Comment (richardlau) [reply lib/links.js:3]: > By tolerating spaces at the start of the line, might this then match the "Fixes:" etc. in patches we float from V8 etc.? (Maybe that's not a problem?)

Not sure. Maybe we need to find example PRs and run `git node metadata` on them to see what they generate? We can do that on both open and closed PRs -- there will be an error about the PR not being landable but the output will still include the generated metadata.

It's entirely possible we may need stricter checking of the right hand side of the colon, but it's hard to speculate on that without more concrete examples.

Comment (richardlau) [reply lib/links.js:3]: Found one. https://github.com/nodejs/node/pull/37276 results in `Fixes: https://github.com/v8:11389` (with and without this PR):
```
$ git node metadata 37276
✔  Done loading data for nodejs/node/pull/37276
----------------------------------- PR info ------------------------------------
Title      deps: V8: workaround for darwin arm64 code page decommit failures (#37276)
   ⚠  Could not retrieve the email or name of the PR author's from user's GitHub profile!
Branch     matinzd:patch-1 -> nodejs:master
Labels     v8 engine, backported-to-v14.x
Commits    2
 - deps: V8: Workaround MacOS 11.2 code page decommit failures
 - bump: v8: v8_embedder_string to 25
Committers 2
 - GitHub <noreply@github.com>
 - matinzd <matin.zadehdolatabad@rambody.me>
------------------------------ Generated metadata ------------------------------
PR-URL: https://github.com/nodejs/node/pull/37276
Fixes: https://github.com/v8:11389
Refs: https://bugs.chromium.org/p/v8/issues/detail?id=11389#c18
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Ash Cripps <acripps@redhat.com>
--------------------------------------------------------------------------------
```

Comment (richardlau) [reply lib/links.js:3]: I've just noticed that for PR-URL we have a different parsing scheme that pulls the links out of the corresponding anchor tag (the parser is parsing HTML). Maybe we should be reusing that logic?

Comment (richardlau) [reply lib/links.js:3]: Reusing the same logic that is used for `refs` and `PR-URL` seems a better solution. Or at least a more consistent one. PR: https://github.com/nodejs/node-core-utils/pull/614

**Comment by richardlau** [comment]
Closing in favour of https://github.com/nodejs/node-core-utils/pull/614

**Pull request closed without merging**